### ISSUE TITLE
Gutenberg / blocks support

### ISSUE THREAD

**Issue by bertiebaggio** [org2blog/org2blog#319]
**tl;dr**: Are Gutenberg blocks likely to be an area which org2blog would devote time to implementing?

----

Thank you for filling out this feature request (FR). Please perform every step and answer every question to the best of your ability. That will make it easier for you and more accessible for us to address the feature request with the highest level of quality in the shortest amount of time.

### Description

Some time ago, Wordpress introduced [Gutenberg](https://wordpress.org/gutenberg/), and with it 'blocks'. At present, org2blog exports its content to Wordpress as a 'classic block', unless I've gravely misunderstood something.

### Prerequisites

* [x] Did you study the [README](https://github.com/org2blog/org2blog/blob/master/README.org) and learn how the existing Org2Blog features work?
* [x] Did you review the issue tracker to see if this FR already
exists or is resolved?
* [x] Have you reviewed the [FUTURE](https://github.com/org2blog/org2blog/blob/master/FUTURE.org) file to see if the FR already exists?

`FUTURE` mentions: 'Update documentation for Gutenberg'. I cannot find other references to this in the repo using GH's search feature

### Good Documentation

Good documentation requires the author to take the user's perspective, address that user, and most importantly, have zero working knowledge of how the software works internally. Developers can rarely do this well for one simple reason: they are too close to their product, so someone else usually has to do it. Typically a Technical Writer is staffed to fill that role, but like most software projects (open source or not), Org2Blog can't quite fit that in the budget this year, so it is up to the developers yet again. However, the good news for all Org2Blog users now is that you are here!

You are already writing this feature request from the viewpoint required for good documentation. You already know how you'll be doing what you want to do with this product: you just need them implemented. Here, my friend is where the magic can happen.

Given your unique take on Org2Blog, how it can improve, the new functionality inspiring you, and your ideal position as a thoughtful user and feature request submitter—everyone is dying to know:

Are you willing to write new or refine the existing documentation as this feature request moves through its lifecycle?

Please choose one of the following:

- [x] Yes.
- [ ] No. Please provide an explanation why.

I am happy to help with this how I can, bearing in mind my at present very limited familiarity with emacs, lisp and org2blog.

### Type of FR

Please try to limit your FR to one type, submit multiple FRs if needed.

Please check the type of feature you are requesting:

- [x] Brand New Feature
- [ ] Change Existing Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Describe the solution you'd like

I would like an option for enabling export of content to Wordpress blocks, instead of a 'classic block.

Why? Wordpress and its plugins are moving in that direction. Using the Gutenberg editor with a classic block is slightly clunky, and some blocks don't function with classic blocks.

As a small but concrete example: I used [GutenTOC](http://tauhidpro.com/gutentoc-advance-table-of-content), which creates a nested table of contents on-the-fly based on heading blocks. A post with headings I created in org and saved with org2blog had a classic block, which GutenTOC didn't see. 

Now, I am probably overlooking an org/org2blog option to generate TOC in this example, but there are others: reusable blocks, containers, image 'diffs' (compare 2 images by dragging a slider).

Naturally it would be churlish to expect org2blog to support all of these straight out of the blocks (forgive the pun), but folks could always load up content in Wordpress for touch ups.

### Describe alternatives you've considered

None, other than deciding not to implement this, which is of course your prerogative! :)

### Screenshots & User Interface Mockups

N/A

### Additional context

This is less of a "do this", more of a "I am happy to spend time investing in learning org2blog and adjusting my workflow, or not if not, no worries". I appreciate that implementing might be rather a large investment! That said, the structure of an org document doesn't feel too different to a (simple) block-based post if you squint just right...

Thanks for your time!

**Comment by grettke** [comment]
Thank you @bertiebaggio for your request. 

My reply is inline.

> tl;dr: Are Gutenberg blocks likely to be an area which org2blog would devote time to implementing?

If users want something then time will get spent on it. 

The hard part is precisely what it means to implement Gutenberg Blocks.

> Some time ago, Wordpress introduced Gutenberg, and with it 'blocks'. At present, org2blog exports its content to Wordpress as a 'classic block', unless I've gravely misunderstood something.

For sake of capturing it here: WordPress stores content has HTML. In Pre-Gutenberg-block Wordpress you directly edited HTML. You had convenience features and plugins but it was more or less straight HTML. In the now Gutenberg-block Wordpress experience you a combination of a GUI Builder and a Rapid Application Development environment. Depending on the user it is cognitively a jarringly difference experience. Fortunately the only thing we have to care about is what Org2Blog _thinks_ about it. 

Org2Blog thinks in Org mode and generates HTML that Wordpress accepts.

If you use Org2Blog today then you are likely using the [https://wordpress.org/plugins/classic-editor/](Classic Editor)

I just created [https://github.com/org2blog/org2blog/issues/322](this ticket) to add that note to the readme.

> FUTURE mentions: 'Update documentation for Gutenberg'. I cannot find other references to this in the repo using GH's search feature

I just added this to that note:

`Note that since Org2Blog uses the Org-Wordpress model there is no integration.`

> I am happy to help with this how I can, bearing in mind my at present very limited familiarity with emacs, lisp and org2blog.

Excellent because I don't know what I would look like to go from Org to Blocks.

> Why? Wordpress and its plugins are moving in that direction. Using the Gutenberg editor with a classic block is slightly clunky, and some blocks don't function with classic blocks.

Don't think of Org2Blog as a plugin though. You are editing Org mode content and just happen to publishing to Wordpress. You can support any Wordpress features that you like with HTML blocks immediately. Figure out how to write it in Wordpress HTML and put use [https://github.com/org2blog/org2blog/#raw-html-creation](Raw HTML Creation). 

To your point I'll update the docs with the expectation that Org2Blog users are probably using the Classic Editor.

> When Org2Blog publishes its posts, because it is one-way, it blows away whatever work you've done with the Block Editor consequently.
> 
> As a small but concrete example: I used GutenTOC, which creates a nested table of contents on-the-fly based on heading blocks. A post with headings I created in org and saved with org2blog had a classic block, which GutenTOC didn't see.

I haven't investigated how Org2Blog works with Blocks because of the one-way thing.

> Now, I am probably overlooking an org/org2blog option to generate TOC in this example, 

It needs to be an Org feature that Org2Blog understands how to convert. If you use a TOC plugin that is a plain old list then it should work perfectly because it is not using Blocks. Not saying you need that, but it would the logical approach.

> Naturally it would be churlish to expect org2blog to support all of these straight out of the blocks (forgive the pun), but folks could always load up content in Wordpress for touch ups.

Is your workflow to write it once in Org mode, publish it, touch it up with the block editor, and never modify and re-publish the Org file again?

> This is less of a "do this", more of a "I am happy to spend time investing in learning org2blog and adjusting my workflow, or not if not, no worries". I appreciate that implementing might be rather a large investment! That said, the structure of an org document doesn't feel too different to a (simple) block-based post if you squint just right...

If you are interested in one-way publishing that would be start. Within that context, what block support would you like added? I use the interactive one-way publishing workflow and do not use Gutenber Blocks even within the raw HTML. Consequently I don't have any perspective or context here: you would need to provide that so we could work through some examples.

**Comment by grettke** [comment]
When you have some ideas on how the feature might look please create a feature request or we can continue the discussion here.

Until then: closing.
